url/urlsearchparams.go

```diff
@@ -0,0 +1,424 @@
+		query = append(query, searchParam{name: k, value: []string{val}})
+	}
+
+	u, _ := url.Parse("")
+	u.RawQuery = query.String()
+	return u
+}
+
+func buildParamsFromArray(r *goja.Runtime, v goja.Value) *url.URL {
+	query := searchParams{}
+
+	o := v.ToObject(r)
+	ex := r.Try(func() {
+		r.ForOf(o, func(val goja.Value) bool {
+			obj := val.ToObject(r)
+
+			var name, value string
+			i := 0
+			// Use ForOf to determine if the object is iterable
+			r.ForOf(obj, func(val goja.Value) bool {
+				if i == 0 {
+					name = fmt.Sprintf("%v", val)
+					i++
+					return true
+				}
+				if i == 1 {
+					value = fmt.Sprintf("%v", val)
+					i++
+					return true
+				}
+				// Array isn't a tuple
+				panic(newInvalidTupleError(r))
+			})
+
+			// Ensure we have two values
+			if i <= 1 {
+				panic(newInvalidTupleError(r))
+			}
+
+			query = append(query, searchParam{
+				name:  name,
+				value: []string{value},
+			})
+
+			return true
+		})
+	})
+
+	if ex != nil {
+		panic(newInvalidTupleError(r))
+	}
+
+	u, _ := url.Parse("")
+	u.RawQuery = query.String()
+	return u
+}
+
+func buildParamsFromMap(r *goja.Runtime, v goja.Value) *url.URL {
+	query := searchParams{}
+	o := v.ToObject(r)
+	ex := r.Try(func() {
+		r.ForOf(o, func(val goja.Value) bool {
+			obj := val.ToObject(r)
+			query = append(query, searchParam{
+				name:  obj.Get("0").String(),
+				value: []string{stringFromValue(r, obj.Get("1"))},
+			})
+			return true
+		})
+	})
+
+	if ex != nil {
+		panic(ex)
+	}
+
+	u, _ := url.Parse("")
+	u.RawQuery = query.String()
+	return u
+}
+
+func stringFromValue(r *goja.Runtime, v goja.Value) string {
+	switch v.ExportType() {
+	case reflectTypeString, reflectTypeInt:
+		return v.String()
+	case reflectTypeArray:
+		vals := []string{}
+		ex := r.Try(func() {
+			r.ForOf(v, func(val goja.Value) bool {
+				vals = append(vals, fmt.Sprintf("%v", val))
+				return true
+			})
+		})
+		if ex != nil {
+			panic(ex)
+		}
+		return strings.Join(vals, ",")
+	case reflectTypeObject:
+		return "[object Object]"
+	default:
+		return fmt.Sprintf("%v", v)
+	}
+}
+
+func createURLSearchParamsPrototype(r *goja.Runtime) *goja.Object {
+	p := r.NewObject()
+
+	p.Set("append", r.ToValue(func(call goja.FunctionCall) goja.Value {
+		if len(call.Arguments) < 2 {
+			panic(newMissingArgsError(r, `The "name" and "value" arguments must be specified`))
+		}
+
+		u := toURL(r, call.This)
+		u.searchParams = append(u.searchParams, searchParam{
+			name:  call.Arguments[0].String(),
+			value: []string{call.Arguments[1].String()},
+		})
+		u.syncSearchParams()
+
+		return goja.Undefined()
+	}))
+
+	p.Set("delete", r.ToValue(func(call goja.FunctionCall) goja.Value {
+		if len(call.Arguments) < 1 {
+			panic(newMissingArgsError(r, `The "name" argument must be specified`))
+		}
+
+		u := toURL(r, call.This)
+		name := call.Arguments[0].String()
+		if len(call.Arguments) > 1 {
+			value := call.Arguments[1].String()
+			arr := searchParams{}
```

Comment:
I've since simplified this approach, but the reason I used approach was to allow for deletion of multiple items in one pass. The Search Params might have have many entries with the name. Which would result in all of them being removed.

Reply:
It still doesn't justify the allocation and the full copy. Even if multiple elements are deleted it's still possible to do it with in-place replacement and with a single pass.